Create basic prismic dom parser

albronils opened the pull request:
- create package.json and add necessary packages for development
- setup prettier + eslint
- create rich-text-parser node-tree and html-serializer to parse prismic rich text into html
- setup test environment with jasmine + babel and create unit tests
- write basic documentation in README.md

Commit:
Create basic prismic dom parser
- create package.json and add necessary packages for development
- setup prettier + eslint
- create rich-text-parser node-tree and html-serializer to parse prismic rich text into html
- setup test environment with jasmine + babel and create unit tests
- write basic documentation in README.md

Commit:
Make HtmlSerializer customisable
- turn static methods into prototype methods
- add options to HTMLSerializer constructor
- RichTextParser can now be passed an HtmlSerializer instance

embrilliant commented on src/html-serializer.js:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  /**
+   * Create <img> wrapped in <p>.
```
An `img` is always wrapped in `p`?

albronils replied:
Yes, this was adopted from the original parser

embrilliant commented on src/html-serializer.js:
```diff
@@ -0,0 +1,188 @@
+      ? ` target="${escapedTarget}" rel="noopener" `
+      : ' '
+
+    return `<a${target}href="${href}">${content}</a>`
```
Would
```
const targetSegment = escapedTarget
      ? `target="${escapedTarget}" rel="noopener"`
      : ''

    return `<a ${targetSegment} href="${href}">${content}</a>`
```
be more in control?

albronils replied:
What do you mean by more in control? It is like this in the prismic-dom lib, but when not providing a target it leaves the a tag with two whitespaces, which seems a bit random to me (let a test fail and was hard to see).

embrilliant replied:
I only mean the ' ' space

Commit:
Improve HtmlSerializer customisation
- add index.js
- export both RichTextParser and serializer in index.js
- Remove default instantiation of HtmlSerializer in RichTextParser to allow smaller bundle size
- Provide better examples/improve documentation

embrilliant commented on README.md:
```diff
@@ -1,2 +1,126 @@
+const htmlSerializer = new HtmlSerializer({
+  imageCopyrightAttribute: 'title'
+})
+
```
Maybe more explanation that the value to the attribute is `element.copyright`? 🤔

embrilliant commented on src/rich-text-parser.js:
```diff
@@ -0,0 +1,25 @@
+   */
+  static parseAsHtml(richText, htmlSerializer) {
+    return htmlSerializer.serializeNodeTree(new NodeTree(richText)).join('')
+  }
```
I wondered why creating an object that only uses the constructor that returns an arr and then join. Why not just a method and also put join there?

Or if using an object, getChildren / getJoinedChildren ..  🤔

albronils replied:
This is for separation of concerns. I think we can argue about the reason for a separate NodeTree class, but mainly this is used as an interface, so that the serializer can know which object to expect.
This way we can also give a user the possibility to provide its own serializer.
A pretty weak reason for this approach, but still valid, is to keep the interface closer to prismic-dom, where you are also exposed a parser with these two methods.

Also the join is for the serialization result not the node tree children

embrilliant replied:
And why not using getChildren in the instance? They (prismic-dom) return an arr directly from the constructor?

embrilliant commented on src/utils/types.js:
```diff
@@ -0,0 +1,78 @@
+export const NODE_TYPES = {
```
Is this a utility?  🤔

albronils replied:
Not really, but it seemed overkill to create a utils/ and a lib/ directory, should I change that? Or create a types/ directory?

embrilliant replied:
Because it feels a bit unorganised. Maybe types/ ?

embrilliant commented on src/node-tree.js:
```diff
@@ -0,0 +1,264 @@
+   * @param {SpanNode[]} spanNodes
+   * @returns {SpanNode[][]}
+   */
+  function groupSpanNodes(spanNodes) {
```
Are there reasons using function declarations instead of named funcs?

Commit:
defaultHyperlinkTarget serializer option | Code Refactoring
- introduce defaultHyperlinkTarget
- apply hyperlink target changes to tests
- make parseAsHtml function clearer
- create getChildren() function on node tree, simplify constructor
- move type files to types/ directory
- rename NODE_TYPES to PRISMIC_ELEMENTS

Commit:
Add .travis.yml

Commit:
Add travis CI build status to README.md

embrilliant commented on src/html-serializer.js:
```diff
@@ -0,0 +1,188 @@
+   */
+  serializeImage(element) {
+    const src = escapeHtml(element.url)
+    const alt = escapeHtml(element.alt || '')
```
I am curious, is it possible that element itself, for example, is undefined? Then it will throw an error.

albronils replied:
No this can't be, the element is a child or sub child of the NodeTree which only contains defined elements, as far as I'm concerned.
Technically if you would use the parser manually and try to break it, passing a malicious node tree or calling this function directly then yes, it would fail. But otherwise we should be fine ^^

embrilliant commented on src/node-tree.js:
```diff
@@ -220,17 +220,21 @@ const processTextBlock = textBlock => {
-    this.children = richText.reduce((children, currentBlock) => {
+    this.richText = richText
+  }
+
```
Hmm...... sorry I thought it would be `this.children = this.getChildren()` and getChildren() can be private

Commit:
Refactor NodeTree child creation
